GPUP: RemoteGraphicsContextGL::reshape does not check for texture size limits

roberto-apple opened the pull request:
#### 9fc408a07d5f53e650be048417525c76dde4848c
<pre>
GPUP: RemoteGraphicsContextGL::reshape does not check for texture size limits
<a href="https://bugs.webkit.org/show_bug.cgi?id=290076">https://bugs.webkit.org/show_bug.cgi?id=290076</a>
<a href="https://rdar.apple.com/146463902">rdar://146463902</a>

Reviewed by Kimmo Kinnunen.

Perform width/height check during GraphicsContextGLANGLE::reshape().
Compute max internal framebuffer size.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLANGLE::reshape):
(WebCore::GraphicsContextGLANGLE::getMaxInternalFrameBufferSize const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeSuccess)):
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeWidthTooLarge)):
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeHeightTooLarge)):

Canonical link: <a href="https://commits.webkit.org/292805@main">https://commits.webkit.org/292805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d4e39e3c49e00d656a66fe351329f3192e6da03

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97186 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102267 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47711 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99231 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74031 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/31240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100189 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12917 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/87887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54378 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12672 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5763 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47108 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5840 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104289 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24261 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83075 "Passed tests") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84014 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82487 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20742 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4719 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17765 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24225 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24048 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27360 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25621 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -171,6 +171,10 @@ static EGLDisplay initializeEGLDisplay(const GraphicsContextGLAttributes& attrs)
     , m_drawingBufferColorSpace(DestinationColorSpace::SRGB())
 {
+    // Set the max internal framebuffer size.
+    int minDimension = std::min(MAX_TEXTURE_SIZE, std::min(MAX_VIEWPORT_DIMS, MAX_RENDERBUFFER_SIZE));
```
grep the webkit whole project to see how these are queried..

kkinnunen-apple replied:
These should be queried  in ANGLE code after the context has been created.

The platform specific part for cocoa should then take the min of this and iosurface max size..

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
```diff
@@ -455,6 +456,7 @@ class WEBCORE_EXPORT GraphicsContextGLANGLE : public GraphicsContextGL {
     uint32_t m_nextExternalSyncName { 0 };
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglImages;
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglSyncs;
+    std::optional<IntSize> m_maxInternalFramebufferSize { std::nullopt };
```
I think you can just store IntSize

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -159,6 +159,15 @@ bool GraphicsContextGLANGLE::initialize()
+    std::array<GCGLint, 2> maxViewportDims { 0, 0 };
+    getIntegerv(GraphicsContextGL::MAX_VIEWPORT_DIMS, maxViewportDims);
+    GCGLint minViewportDim = std::min(maxViewportDims[0], maxViewportDims[1]);
+    GCGLint minDimension = std::min(maxTextureSize, std::min(maxRenderBufferSize, minViewportDim));
```
I think calculating minViewportDim like that isn't what's intended.

```
m_maxInternalFramebufferSize = {maxViewportDims[0], maxViewportDims[1]};
m_maxInternalFramebufferSize.clampToMinimumSize({maxTextureSize, maxTextureSize});
m_maxInternalFramebufferSize.clampToMinimumSize({maxRenderbufferSize, maxRenderbufferSize});
```

roberto-apple replied:
Added correct computation

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
```diff
@@ -455,6 +456,7 @@ class WEBCORE_EXPORT GraphicsContextGLANGLE : public GraphicsContextGL {
     uint32_t m_nextExternalSyncName { 0 };
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglImages;
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglSyncs;
+    IntSize m_maxInternalFramebufferSize { 0, 0 };
```
{ 0, 0} is redundant

roberto-apple replied:
Removed

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -308,6 +308,9 @@ static EGLDisplay initializeEGLDisplay(const GraphicsContextGLAttributes& attrs)
 {
+    // Compute platform-specific max internal framebuffer size.
+    GCGLint minDimension = std::min(m_maxInternalFramebufferSize.minDimension(), IOSurface::maximumSize().minDimension());
+    m_maxInternalFramebufferSize = IntSize(minDimension, minDimension);
```
```
m_maxInternalFramebufferSize .clampToMinimumSize(IOSurface::maximumSize());
```

roberto-apple replied:
Added correct computation

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -3325,6 +3337,11 @@ void GraphicsContextGLANGLE::addError(GCGLErrorCode errorCode)
     m_errors.add(errorCode);
 }
 
+IntSize GraphicsContextGLANGLE::getMaxInternalFrameBufferSize() const
```
Maybe it's simpler to not add this one, and instead use impossibly large size as the size in tests.

roberto-apple replied:
Removed

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -159,6 +159,15 @@ bool GraphicsContextGLANGLE::initialize()
 
+    // Compute initial platform-independent max internal framebuffer size.
+    GCGLint maxTextureSize = getInteger(GraphicsContextGL::MAX_TEXTURE_SIZE);
+    GCGLint maxRenderBufferSize = getInteger(GraphicsContextGL::MAX_RENDERBUFFER_SIZE);
```
It's "renderbuffer", not "render buffer", (MAX_RENDERBUFFER instead of MAX_RENDER_BUFFER), so the variable should be maxRenderbufferSize

roberto-apple replied:
Updated name

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
```diff
@@ -455,6 +455,7 @@ class WEBCORE_EXPORT GraphicsContextGLANGLE : public GraphicsContextGL {
     uint32_t m_nextExternalSyncName { 0 };
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglImages;
     UncheckedKeyHashMap<uint32_t, void*, IntHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_eglSyncs;
+    IntSize m_maxInternalFramebufferSize { };
```
No need to explicitly initialise, it is initialised by the default constructor.

roberto-apple replied:
Removed initialization braces

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -159,6 +159,15 @@ bool GraphicsContextGLANGLE::initialize()
     GL_ClearColor(0, 0, 0, 0);
 
+    // Compute initial platform-independent max internal framebuffer size.
+    GCGLint maxTextureSize = getInteger(GraphicsContextGL::MAX_TEXTURE_SIZE);
```
I forgot to mention:
Like the above GL_ClearColor, here you should call the internal implementation and not the public API.
So these would be GL_GetIntegerv

roberto-apple replied:
Replaced with calls to GL_GetIntegerv.

kkinnunen-apple reviewed (approved):
With the small change

roberto-apple committed:
GPUP: RemoteGraphicsContextGL::reshape does not check for texture size limits
https://bugs.webkit.org/show_bug.cgi?id=290076
rdar://146463902

Reviewed by Kimmo Kinnunen.

Perform width/height check during GraphicsContextGLANGLE::reshape().
Compute max internal framebuffer size.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLANGLE::reshape):
(WebCore::GraphicsContextGLANGLE::getMaxInternalFrameBufferSize const):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeSuccess)):
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeWidthTooLarge)):
(TestWebKitAPI::TEST_F(GraphicsContextGLCocoaReshapeTest, reshapeHeightTooLarge)):

Canonical link: https://commits.webkit.org/292805@main
